src/SignatureTM.cpp

```diff
@@ -129,37 +128,51 @@ void SignatureTM::validate() const
         const RevocationValuesType &t = uProps->unsignedSignatureProperties()->revocationValues().front();
         if(!t.oCSPValues().present())
             THROW_MAIN(exception, "OCSPValues is missing");
-        if(t.oCSPValues()->encapsulatedOCSPValue().size() != 1)
-            THROW_MAIN(exception, "More than one EncapsulatedOCSPValue object is not supported");
-
-        const OCSPValuesType::EncapsulatedOCSPValueType &resp = t.oCSPValues()->encapsulatedOCSPValue().front();
-        OCSP ocsp(vector<unsigned char>(resp.data(), resp.data()+resp.size()));
-        try {
-            ocsp.verifyResponse(signingCertificate());
-        } catch(const Exception &e) {
-            exception.addCause(e);
-        }
-
-        struct tm producedAt = ASN1TimeToTM(ocsp.producedAt());
-        time_t producedAt_t = mktime(&producedAt);
-        if(!X509CertStore::instance()->verify(ocsp.responderCert(), &producedAt_t))
-            EXCEPTION_ADD(exception, "Unable to verify responder certificate");
 
-        if(profile().find(BDoc::ASIC_TM_PROFILE) != string::npos)
+        /*
+         * Find OCSP response that matches with signingCertificate.
+         * If none is found throw all OCSP validation exceptions.
+         */
+        bool foundSignerOCSP = false;
+        vector<Exception> ocspExceptions;
+        for(const OCSPValuesType::EncapsulatedOCSPValueType &resp: t.oCSPValues()->encapsulatedOCSPValue())
         {
-            string method = Digest::digestInfoUri(ocsp.nonce());
-            if(method.empty())
-                THROW("Nonce digest method is missing");
-            Digest calc(method);
-            calc.update(getSignatureValue());
-            vector<unsigned char> digest = calc.result();
-            vector<unsigned char> respDigest = Digest::digestInfoDigest(ocsp.nonce());
-            if(digest != respDigest)
+            OCSP ocsp(vector<unsigned char>(resp.data(), resp.data()+resp.size()));
+            try {
+                ocsp.verifyResponse(signingCertificate());
+                foundSignerOCSP = true;
+            } catch(const Exception &e) {
+                ocspExceptions.push_back(e);
+                continue;
+            }
+
+            struct tm producedAt = ASN1TimeToTM(ocsp.producedAt());
```

Comment:
Will this (until line 169) execute for every response or only for the one matching the signing certificate?
If only for one, for clarity, it would make sense to keep it outside the loop